### PR TITLE
Disallow isolated nodes for Eulerian Path

### LINKED ISSUE #5594: Inconsistent implementation of Euler algorithms
According to my research, having an Euler Circuit and/or Euler Path does not require a graph to be connected. Instead, it is enough that all of its vertices **with nonzero degree** belong to a single connected component. Nevertheless, there are some resources which states that a graph can have Euler Path and/or Euler Circuit  iff it is connected.

In the doc_string, it is stated that `has_eulerian_path` method follows the first opinion. 
```
A directed graph has an Eulerian path iff:
- at most one vertex has out_degree - in_degree = 1,
- at most one vertex has in_degree - out_degree = 1,
- every other vertex has equal in_degree and out_degree,
- and all of its vertices with nonzero degree belong to a single connected component of the underlying undirected graph.

An undirected graph has an Eulerian path iff:
- exactly zero or two vertices have odd degree,
- and all of its vertices with nonzero degree belong to a single connected component.
```

However, when we look at the source code, we see that it weirdly requires connectedness for undirected graphs but not for directed ones. Here is the related part:

```
  if G.is_directed():
      # Remove isolated nodes (if any) without altering the input graph
      nodes_remove = [v for v in G if G.in_degree[v] == 0 and G.out_degree[v] == 0]
      if nodes_remove:
          G = G.copy()
          G.remove_nodes_from(nodes_remove)
```
If graph is directed, it removes the isolated nodes. But at the last part, it checks for connectedness both for directed (already removed isolated ones) and undirected graphs.

```
return (
            unbalanced_ins <= 1 and unbalanced_outs <= 1 and nx.is_weakly_connected(G)
        )
else:
    # We know it is not eulerian, so degree of source must be odd.
    if source is not None and G.degree[source] % 2 != 1:
        return False

    # Sum is 2 since we know it is not eulerian (which implies sum is 0)
    return sum(d % 2 == 1 for v, d in G.degree()) == 2 and nx.is_connected(G)
```
As a result, following examples return different results:

```
>>> G = nx.DiGraph([(0, 1), (1, 2), (2, 0)])
>>> G.add_node(3)
>>> nx.draw(G)
>>> nx.has_eulerian_path(G)
True
```
image
```
>>> G = nx.Graph([(0, 1), (1, 2), (0, 2)])
>>> G.add_node(3)
>>> nx.draw(G)
>>> nx.has_eulerian_path(G)
False
```
image

On the other hand, `is_eulerian` method requires connectedness for both graph types. IMO, there is an incosistency in the source code. If the first approach is to be accepted, it should be implemented both for Euler Paths and Euler Circuits since an Euler Circuit is nothing but an Euler Path that starts and ends at the same vertex. If you think that connectedness must be a condition, it should be implemented for both directed and undirected graph types.

@rossbar @dschult @MridulS

### PR DESCRIPTION
**Pull request by dtekinoglu**
Fix #5594.

### CODE REVIEW

File: networkx/algorithms/tests/test_euler.py [hunk 1]
```diff
@@ -140,16 +140,6 @@ def test_has_eulerian_path_directed_graph(self):
...
-    def test_has_eulerian_path_isolated_node(self):
-        # Test directed graphs without isolated node returns True
-        G = nx.DiGraph()
-        G.add_edges_from([(0, 1), (1, 2), (2, 0)])
-        assert nx.has_eulerian_path(G)
-
-        # Test directed graphs with isolated node returns True
-        G.add_node(3)
-        assert nx.has_eulerian_path(G)
-
```
Comment (dschult) [inline networkx/algorithms/tests/test_euler.py:152]: Let's leave the test code in but change the result tested for with the isolated node to `returns False` and then `assert not ...`.   I think the name of the function no longer makes sense either. It is tempting to just include these in the previous test function -- that is, just take out the function signature line. At least the name of that function describes what we are testing. Some people like to have a separate function for each test, so if you can come up with a good name for it, that is a reasonable approach too.  My desire is just to have the tests to ensure that we know what we want to happen.

Thanks for this!  Nice work...

File: networkx/algorithms/euler.py [hunk 2]
```diff
@@ -37,10 +42,18 @@ def is_eulerian(G):
...
+    you can first remove such vertices and then call `is_eulerian` as below example shows.
+
+    >>> G = nx.Graph([(0, 1), (1, 2), (0, 2)])
+    >>> G.add_node(3)
+    >>> nx.is_eulerian(G)
+    False
+
+    >>> G.remove_nodes_from(list(nx.isolates(G)))
+    >>> nx.is_eulerian(G)
+    True
```
Comment (rossbar) [inline networkx/algorithms/euler.py:55]: This example does a great job of highlighting the behavior with isolated nodes. There might be a way to keep things clear but use less text, maybe something like:

```suggestion
    Graphs with isolated nodes are not Eulerian:

    >>> G = nx.Graph([(0, 1), (1, 2), (0, 2)])
    >>> G.add_node(3)
    >>> nx.is_eulerian(G)
    False
    
    Isolated nodes can be removed to evaluate whether the
    remaining graph is Eulerian:

    >>> G.remove_nodes_from(list(nx.isolates(G)))
    >>> nx.is_eulerian(G)
    True
```

This is really subjective and just a suggestion - the original wording is totally fine so feel free to ignore!

File: networkx/algorithms/euler.py [hunk 3]
```diff
@@ -224,8 +242,8 @@ def has_eulerian_path(G, source=None):
         - at most one vertex has out_degree - in_degree = 1,
         - at most one vertex has in_degree - out_degree = 1,
         - every other vertex has equal in_degree and out_degree,
-        - and all of its vertices with nonzero degree belong to a
-          single connected component of the underlying undirected graph.
+        - and all of its vertices belong to a single connected
+        component of the underlying undirected graph.
```
Comment (rossbar) [inline networkx/algorithms/euler.py:246]:
```suggestion
        - and all of its vertices belong to a single connected
          component of the underlying undirected graph.
```

I think this is necessary to get the list to render correctly in the generated html!

File: networkx/algorithms/euler.py [hunk 4]
```diff
@@ -253,6 +274,20 @@ def has_eulerian_path(G, source=None):
     -------
     Bool : True if G has an Eulerian path.
 
+    Example
```
Comment (rossbar) [inline networkx/algorithms/euler.py:277]: If you do end up accepting the wording suggestion for the other example, similar wording should be applied here for consistency.

**Review by rossbar — APPROVED** [review]
This LGTM and covers all the suggested changes in https://github.com/networkx/networkx/issues/5594#issuecomment-1120213701. Just a couple minor wording/formatting suggestions but they're not blockers. Thanks @dtekinoglu !

**Review by jarrodmillman — APPROVED** [review]

**Pull request merged**